content/blog/2025-05-13-flat-bril/index.md

```diff
@@ -0,0 +1,170 @@
+[[extra.authors]]
+name = "Samuel Breckenridge"
++++
+
+Typically, the way to implement an interpreter is to create an explicit AST data type, but this requires allocating AST nodes on the heap. Instead, as demonstrated in [this blogpost](https://www.cs.cornell.edu/~asampson/blog/flattening.html) by Adrian, we can flatten the AST into an array, i.e. pack all AST nodes into one single contiguous array, and refer to children in the AST using array indices (as opposed to pointers). This has enormous performance benefits!
+
+Since Bril is based on commands rather than expressions, there are no ASTs to flatten. However, in the existing Rust interpreter, programs, functions and instructions are all represented using structs that contain pointers to the heap. For instance this is the representation of a value instruction:
+
+```Rust
+Value {
+        args: Vec<String>,
+        dest: String,
+        funcs: Vec<String>,
+        labels: Vec<String>,
+        op: ValueOps,
+        pos: Option<Position>,
+        op_type: Type,
+    }
+```
+
+`Vec`s and `String`s in Rust are both implemented using pointers, so the representation is clearly not flat even before considering that functions contain pointers to instructions and programs contain pointers to functions. Our goal was to adopt the same approach of flattening data structures and packing them into a single contiguous array for Bril. The resulting representation of Bril programs is then a flat file containing these arrays that can be directly mapped into memory and interpreted, analogous to the approach laid out in [another blogpost](https://www.cs.cornell.edu/~asampson/blog/flatgfa.html) by Adrian. This requires two main components:
+
+1. Infrastructure to convert existing Bril JSON files to/from our flattened format
+2. An alternate Bril interpreter that operates directly on the flattened data structure (as opposed to the existing one brili, which has to parse JSON)
+
+Our complete implementation can be found at https://github.com/ngernest/flat-bril/tree/main
+
+## Design
+
+### Flat data structures
+
+A key design decision was what the flattened data structures would look like. We clearly need similar fields to those of the original representation, but we need a strategy for flattening `Vec`s and `String`s and other non-flat types. Instead of storing variables, labels and function names as Strings and Vec<String>s we aggregate all the names referenced throughout a function into three contiguous arrays of bytes that are stored in our function representation. Then instead of using Strings, our instruction representation stores pairs of indices that can be used to extract the relevant name from the function-level contiguous array. One subtlety here is that if we want to use a Vec, we cannot just index into the array containing all the names, we would not know where the boundaries between each name are! We need to use an intermediary array that stores indices into the top-level name array.
```

Comment:
You probably want some more code backticks when you refer to Rust types. For example:

> as Strings and Vec<String>s

->

> as `Strings` and `Vec<String>`s